File: mindflow/generate_index.py
"""
Generate an index for a list of resolved paths.
"""

from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor, as_completed
from mindflow.requests.unindexed_reference import request_unindexed_references
from mindflow.requests.index_references import request_index_references
from mindflow.resolve.resolvers.base_resolver import Resolved

PACKET_SIZE = 2 * 1024 * 1024

def get_size(resolved_path):
    """
    Get the size of a file.
    """
    return resolved_path.size_bytes

def process_packet(packet):
    """
    Process a packet of files.
    """
    result = []
    for resolved_path in packet:
        reference = resolved_path.create_reference()
        if reference:
            result.append(reference)

    return result

def generate_index(resolved_paths: list[Resolved]):
    """
    Generate an index for a list of resolved paths.
    """
    processed_hashes = []

    with ProcessPoolExecutor(max_workers=50) as executor:
        sizes = list(executor.map(get_size, resolved_paths))

    packets = []
    packet = []

    size = 0
    total_size = 0

    for file_path, file_size in zip(resolved_paths, sizes):
        if size + file_size <= PACKET_SIZE:
            packet.append(file_path)
            size += file_size
        else:
            packets.append(packet)
            packet = [file_path]
            total_size += size
            size = file_size

    if packet:
        total_size += size
        packets.append(packet)

    # Create a thread pool with 4 worker threads
    with ThreadPoolExecutor(max_workers=500) as executor:
        # Submit the tasks to the thread pool
        results = [executor.submit(process_packet, packet) for packet in packets]

        # Process the results as they complete
        processed_hashes = []
        for f in as_completed(results):
            result = f.result()
            hashes = [reference.hash for reference in result]
            unindexed_hashes = request_unindexed_references(hashes)
            request_index_references({r.hash: r for r in result}, unindexed_hashes)
            processed_hashes.extend(hashes)

    print(f'Total content size: MB {total_size / 1024 / 1024:.2f}')
    return hashes

File: mindflow/src/utils/generate_index.rs
use std::collections::{HashMap};
use indicatif::{ProgressBar, ProgressStyle};

use rayon::prelude::*;

use reqwest::Client;
use tokio::task::spawn_blocking;

use crate::requests::unindexed_references::{request_unindexed_references};
use crate::requests::index_references::{request_index_references};
use crate::resolve::file_path_resolver::ResolvedFilePath;
use crate::resolve::resolver_trait::Resolved;
use crate::utils::reference::Reference;

const PACKET_SIZE: u64 = 2 * 1024 * 1024;

pub async fn generate_index(resolved_paths: Vec<ResolvedFilePath>) -> Vec<String> {
    let client = Client::new();
    let mut processed_hashes: Vec<String> = Vec::new();
    
    let sizes: Vec<u64> = resolved_paths
        .par_iter()
        .filter_map(|resolved_path| {
            resolved_path.size_bytes()
        })
        .collect();

    let mut packets: Vec<Vec<ResolvedFilePath>> = Vec::new();
    let mut packet: Vec<ResolvedFilePath> = Vec::new();

    let mut packet_size: u64 = 0;
    let mut total_size = 0;

    for (resolved_path, resolved_size) in resolved_paths.into_iter().zip(sizes) {
        if resolved_size > PACKET_SIZE {
            println!("LARGE FILE: {}", resolved_path.path);
            continue
        }
        if packet_size + resolved_size <= PACKET_SIZE {
            packet.push(resolved_path);
            packet_size += resolved_size;
        } else {
            packets.push(packet);
            packet = vec![resolved_path];
            total_size += packet_size;
            packet_size = resolved_size;
        }
    }

    if !packet.is_empty() {
        total_size += packet_size;
        packets.push(packet);
    }

    let pb = ProgressBar::new(packets.len() as u64);
    pb.set_style(ProgressStyle::default_bar());
    pb.set_message("Indexing files");
    let mut packet_index = 0;

    for packet in packets {
        let result = spawn_blocking( move || {
            let reference_vec: Vec<Reference> = packet
                .par_iter()
                .filter_map(|resolved_path| {
                    resolved_path.create_reference()
                })
                .collect();
            reference_vec
        }).await.unwrap();

        let mut hashes: Vec<String> = Vec::new();
        let mut data_map: HashMap<String, Reference> = HashMap::new();

        for reference in result {
            hashes.push(reference.hash.clone());
            data_map.insert(reference.hash.clone(), reference);
        }

        let unindexed_reference_response = request_unindexed_references(&client, &hashes).await;
        let unindexed_hashes = match unindexed_reference_response {
            Ok(unindexed_reference_response) => unindexed_reference_response.unindexed_hashes,
            Err(e) => {
                println!("Error: Could not get unindexed hashes: {}", e);
                Vec::new()
            }
        };
        if unindexed_hashes.len() != 0 {
            request_index_references(&client, data_map, unindexed_hashes).await;
        }

        packet_index += 1;
        pb.set_position(packet_index);


        processed_hashes.extend(hashes);
    }
    println!("Total content size: MB {}\n", format!("{:.2}", total_size as f64 / 1024.0 / 1024.0));
    processed_hashes
}